File: Orchestration/fastapi/app/security/signing.py
import hashlib
import hmac
from dataclasses import dataclass
from uuid import UUID

SIGNATURE_VERSION = "v1"
KEY_VERSION_HEADER = "X-CatalogGuard-Key-Version"
SERVICE_HEADER = "X-CatalogGuard-Service"
TIMESTAMP_HEADER = "X-CatalogGuard-Timestamp"
NONCE_HEADER = "X-CatalogGuard-Nonce"
SIGNATURE_HEADER = "X-CatalogGuard-Signature"


@dataclass(frozen=True, slots=True)
class SignedHeaders:
    key_version: str
    service_id: str
    timestamp: int
    nonce: UUID
    signature: str

    def as_http_headers(self) -> dict[str, str]:
        return {
            KEY_VERSION_HEADER: self.key_version,
            SERVICE_HEADER: self.service_id,
            TIMESTAMP_HEADER: str(self.timestamp),
            NONCE_HEADER: str(self.nonce),
            SIGNATURE_HEADER: self.signature,
        }


def body_sha256(body: bytes) -> str:
    return hashlib.sha256(body).hexdigest()


def canonical_message(
    *,
    key_version: str,
    service_id: str,
    timestamp: int,
    nonce: UUID,
    method: str,
    path: str,
    body: bytes,
) -> bytes:
    components = (
        SIGNATURE_VERSION,
        key_version,
        service_id,
        str(timestamp),
        str(nonce),
        method.upper(),
        path,
        body_sha256(body),
    )
    return "\n".join(components).encode("utf-8")


def sign_http_message(
    *,
    secret: str,
    key_version: str,
    service_id: str,
    timestamp: int,
    nonce: UUID,
    method: str,
    path: str,
    body: bytes,
) -> SignedHeaders:
    signature = hmac.new(
        secret.encode("utf-8"),
        canonical_message(
            service_id=service_id,
            key_version=key_version,
            timestamp=timestamp,
            nonce=nonce,
            method=method,
            path=path,
            body=body,
        ),
        hashlib.sha256,
    ).hexdigest()
    return SignedHeaders(
        key_version=key_version,
        service_id=service_id,
        timestamp=timestamp,
        nonce=nonce,
        signature=signature,
    )

File: Web/server/integrations/orchestration/signing.ts
import "server-only";

import { createHash, createHmac, timingSafeEqual } from "node:crypto";
import { z } from "zod";

import { AppError } from "@/lib/contracts/errors";

export const orchestrationSigningHeaders = {
  keyVersion: "x-catalogguard-key-version",
  service: "x-catalogguard-service",
  timestamp: "x-catalogguard-timestamp",
  nonce: "x-catalogguard-nonce",
  signature: "x-catalogguard-signature",
} as const;

type SigningInput = {
  secret: string;
  keyVersion: string;
  serviceId: string;
  timestamp: number;
  nonce: string;
  method: string;
  path: string;
  body: Uint8Array;
};

function bodySha256(body: Uint8Array): string {
  return createHash("sha256").update(body).digest("hex");
}

export function canonicalOrchestrationMessage(input: Omit<SigningInput, "secret">): Buffer {
  return Buffer.from([
    "v1",
    input.keyVersion,
    input.serviceId,
    String(input.timestamp),
    input.nonce,
    input.method.toUpperCase(),
    input.path,
    bodySha256(input.body),
  ].join("\n"), "utf8");
}

export function signOrchestrationMessage(input: SigningInput): string {
  return createHmac("sha256", input.secret)
    .update(canonicalOrchestrationMessage(input))
    .digest("hex");
}

export type VerifiedOrchestrationMessage = {
  keyVersion: string;
  serviceId: string;
  timestamp: number;
  nonce: string;
};

export function verifyOrchestrationMessage(input: {
  headers: Headers;
  body: Uint8Array;
  method: string;
  path: string;
  expectedKeyVersion: string;
  expectedServiceId: string;
  signingSecret: string;
  maxClockSkewSeconds: number;
  now?: number;
}): VerifiedOrchestrationMessage {
  const keyVersion = input.headers.get(orchestrationSigningHeaders.keyVersion);
  const serviceId = input.headers.get(orchestrationSigningHeaders.service);
  const timestampHeader = input.headers.get(orchestrationSigningHeaders.timestamp);
  const nonce = input.headers.get(orchestrationSigningHeaders.nonce);
  const signature = input.headers.get(orchestrationSigningHeaders.signature);
  if (!keyVersion || !serviceId || !timestampHeader || !nonce || !signature) {
    throw new AppError({ code: "SERVICE_AUTHENTICATION_FAILED", message: "Service authentication failed.", status: 401 });
  }
  const timestamp = Number(timestampHeader);
  if (!Number.isSafeInteger(timestamp) || !z.uuid().safeParse(nonce).success || keyVersion !== input.expectedKeyVersion || serviceId !== input.expectedServiceId || !/^[a-f\d]{64}$/i.test(signature)) {
    throw new AppError({ code: "SERVICE_AUTHENTICATION_FAILED", message: "Service authentication failed.", status: 401 });
  }
  if (Math.abs((input.now ?? Math.floor(Date.now() / 1000)) - timestamp) > input.maxClockSkewSeconds) {
    throw new AppError({ code: "SERVICE_MESSAGE_STALE", message: "The service message is outside the accepted time window.", status: 401 });
  }
  const expected = signOrchestrationMessage({ secret: input.signingSecret, keyVersion, serviceId, timestamp, nonce, method: input.method, path: input.path, body: input.body });
  const supplied = Buffer.from(signature, "hex");
  const calculated = Buffer.from(expected, "hex");
  if (supplied.length !== calculated.length || !timingSafeEqual(supplied, calculated)) {
    throw new AppError({ code: "SERVICE_AUTHENTICATION_FAILED", message: "Service authentication failed.", status: 401 });
  }
  return { keyVersion, serviceId, timestamp, nonce };
}
